### ISSUE TITLE
Allow Sharing from the Desktop

### ISSUE THREAD

**Issue by MTRichards** [owncloud/client#1444]
As a user, I want to be able to click on a file, and select share with ownCloud, and then be able to share a file with someone as a link, or as a file shared directly with them in ownCloud using the share API.

This is partly a deeper integration into ownCloud and the desktop OS to enable the menu, and partly an implementation of the sharing API on the desktop.  Likely there are two steps:
1) share with link (share, unshare)
2) share with user of the system (share, unshare, set CRUDS permissions)

**Comment by dragotin** [comment]
@rullzer would you mind to create a submit request against owncloud/client master branch? That would be nice.

**Comment by dragotin** [comment]
I merged the branch of @rullzer today (thanks!) and adopted the Nautilus plugin accordingly, so that sharing from Linux Nautilus is possible now.

Remaining tasks:
* [x] SocketAPI share commando
* [ ] Windows Explorer plugin adopted
* [ ] Mac Finder plugin adopted
* [x] Nautilus plugin adopted
* [ ] KDE Dolphin plugin adopted (not release critical)

**Comment by rullzer** [comment]
Public url sharing should be working pretty OK in that merge.

However sharing with users and groups (and modifying access) is still under development. I'll hopefully submit a patch for that soon. Then somebody can play with making it look decent.

**Comment by rullzer** [comment]
I just created another pull request https://github.com/owncloud/client/pull/2700
This allows user/group sharing. Still not complete. But maybe somebody can have a look at the UI already :)

**Comment by jancborchardt** [comment]
@rullzer awesome! Some feedback regarding the sharing dialog:

* [ ] It needs to be clearer what will be shared. On the top of the dialog there should ideally be a thumbnail and the filename (or folder name, or multiple file names).
* [ ] Merge the user/group field into one and label it »User or group …«. The label above should only be »Share with«
* [ ] We need type-ahead autocompletion for the user/group. Otherwise you need to know the exact username. A server-API is needed for that (cc @DeepDiver1975) – and we shouldn’t offer user/group sharing until we have it since we can’t expect people to know usernames of people.

**Comment by alefattorini** [comment]
@jancborchardt can I help? How can I test this dev version on my pc?

**Comment by jancborchardt** [comment]
@rullzer @dragotin @guruz basically what I’m saying is that maybe it’s wiser to start with only the »Share by link« feature in the dialog, and postpone the sharing to other users. Until we have the proper API to return users from the server.

**Comment by Bridgesltd** [comment]
completely agree with @jancborchardt comments above. It would likely be easier to implement the whole sharing feature by keeping it simple first and then expanding on to users / groups.

**Comment by rullzer** [comment]
@jancborchardt @Bridgesltd That was my idea as well initially. But then I went ahead and started implementing the rest as well. So I guess the best way for now is to just go ahead and remove all the user/group stuff. At least from the UI.

So we indeed need to get the server involved. Since right now we need two lines (one to share with group and one to share with user). Since they are different share types. So we do not get an auto complete list from the server we will not know what type of share to create.

I'll branch of my current work and create a version that only shares a link. Hopefully there will be yet another pull request soon ;)

P.s. I rebase all my branches on master. However I made a pull request to the Sharedialog branch. So that most likely makes merging less trivial...

**Comment by jancborchardt** [comment]
@rullzer cool, great stuff! :) Yeah, sometimes we have to see what’s feasible to include in a next release. 

We for sure want to include the sharing with users/groups in the future, but we need to have it polished for that. :)

**Comment by rullzer** [comment]
Ok so more commits are done :).

- [x] I removed the user/group sharing for now.
- [x] Added top section that shows what is being shared. I can't really test the icon since I do not have icons on WM. So somebody please test.
- [x] Cleaned up the code a bit.
- [ ] More error handling. A lot is already done but still this has to be fixed further. It is a bit tough to test sometimes tough.

So please test and review the branch.

**Comment by dragotin** [comment]
@rullzer for tests, I use netcat, nc, ie. ```nc -U /run/user/1000/ownCloud/socket``` which opens the socket API, and that allows to send commands to it via commandline. Very convenient.

**Comment by dragotin** [comment]
Also to be considered here: https://github.com/owncloud/client/issues/2732

**Comment by dragotin** [comment]
I today committed 95f7e83c5c567030ddda1052db0332d970fb5dbc to finish the socket API for sharing. Also commit 82e5e36c57d2451de27c151afd85f296a12da5f2 for the natuilus plugin.

So the current picture is:
- [x] Socket API implemented for sharing (rullzer, dragotin)
- [x] Additions to Nautilus plugins doing context menu (dragotin)
- [ ] Additions to the MacOSX Finder plugin to add a context menu (dragotin, guruz)
- [ ] Additions to the Windows Explorer plugin to add a context menu (danimo)

**Comment by guruz** [comment]
Finished the OS X Finder integration and tested on 10.9.
@lmaestro Could you check 10.10 ?
It's just about the share dialog showing up.. and Finder not crashing ;)

**Comment by lmaestro** [comment]
I have just tested the latest build (1928) on OS X 10.10 and it doesn't crash.

**Comment by dragotin** [comment]
The base functionality is done, so I think this entry can be closed. 

Missing parts and/or bugs are tracked in more specific issues.

**Comment by phil-davis** [comment]
Sorry for poking this old thread, but I can't see what happened to the future of being able to share with a user/group from the desktop client "Share with ownCloud/ownCloud Sharing" dialog.
Is that still coming?
If so, can somebody put a pointer here to the relevant issue so it will be easy for others to find their way.

**Comment by rullzer** [comment]
@phil-davis yes it will come. But we first need proper server side support. The plan is to get that in for 8.2

**Comment by phil-davis** [comment]
Thanks for the quick response. It might be easy for others to find if there is an open issue on client that documents the need/requirement and mentions that it is awaiting some server-side implementation first.
When there is any moderately-functional bare bones of server and client to test I am very happy to help with debugging.

**Comment by elgs** [comment]
I installed ownCloud client from brew cask, but it seems the share options is still missing from the context menu. Am I missing anything?

**Comment by Carharttguy** [comment]
@elgs: 2 comments above you @rullzer states that it will come when there is proper server side supprt. This will be the case for ownCloud 8.2. Now ownCloud latest version is 8.1.

**Comment by ziliboban** [comment]
Hi All!
We've installed ownCloud 8.2 but it seems that sharing context menu is missing in OS X. Is it supported as planned in 8.2 or not? 
i'm using OS X client 2.0.2 (build 2818).

**Comment by phil-davis** [comment]
You can share a link (with or without password) using the current client 2.0.*
The code for the client to be able to share with another user (of the ownCloud server) is coming.

**Comment by ziliboban** [comment]
@phil-davis it is strange but i don't see any sharing options when in the context menu (see img, sorry it is in russian) 
![2015-10-29 15 48 10](https://cloud.githubusercontent.com/assets/15379310/10818696/a69f46c8-7e54-11e5-836b-7a96fda840e7.png)

**Comment by phil-davis** [comment]
My Nepali and Greek is better than my Russian :)
There should be an entry "Share with ownCloud". I am not aware of that being broken in OS X or Russian - but who knows!

**Comment by ziliboban** [comment]
@phil-davis thanks! :) 
Maybe i need to try to switch to english, will try it.

**Comment by Kernald** [comment]
I just checked on the french version, it works fine. The ownCloud entry is just below the send to trash one.

**Comment by ziliboban** [comment]
With english it is the same, maybe i should reinstall the client.
![screen shot 2015-10-29 at 16 01 44](https://cloud.githubusercontent.com/assets/15379310/10818981/8da20d0c-7e56-11e5-897d-573b4bc31326.png)

**Comment by ziliboban** [comment]
Reinstalled the client and now it works. 
Thank you guys :+1:

**Comment by dragotin** [comment]
@ziliboban maybe that plugin was not enabled before? Since 10.10 IIRC it can be en- and disabled.

Thanks for letting us know that it works now.

**Comment by elgs** [comment]
It seems it still doesn't work for me. I use `brew cask remove owncloud` and `brew cask install owncloud`.

**Comment by ziliboban** [comment]
To make it work i did following:
- swiched to English language in OS X
- downloaded client installation package again
- ran installation process and it actually uninstalled existing client
- ran installation process again and installed client
- checked that it works in English
- swiched back to Russian

**Comment by McCovican** [comment]
Apologies if I've missed something along the way, but is this available properly for Windows (7)? I made sure to uninstall and reinstall the most recent version, but I still have no right-click options in windows explorer. Filetypesman/Filemenutools don't appear to show any entries from OwnCloud.

**Comment by tarator** [comment]
This feature would be nice to work also with KDE's Dolphin.

**Comment by elgs** [comment]
Strangely it still doesn't work for me. I installed the client using homebrew.

**Comment by danimo** [comment]
@tarator It will with KF 5.16 and KDE Application 15.12. Current versions lack API to implement this.

**Comment by tarator** [comment]
@danimo Thanks! I'm looking forward to get that.

**Comment by herrdeh** [comment]
Would love this feature for nautilus and nemo at debian and ubuntu. What's the status?

**Comment by dragotin** [comment]
It is there for nautilus and nemo. Install the packages as described here https://software.opensuse.org/download/package?project=isv:ownCloud:desktop&package=owncloud-client and enjoy.

**Comment by herrdeh** [comment]
Thanks, dragotin. But I never could find any handle for it, neither in nemo nor nautilus. I'm on ubuntu 16.10,  nemo 3.0.6 and nautilus 3.20.3.
I installed the sync client as suggested, in addition I have the "owncloud-client-nemo" and "owncloud-client.-nautilus" package.

I expected to right-click a file or folder and find something like "share..." - which will provide the OC URL.
Yes - I tried it with files which are synced to OC. (-;
